calico_rkt/calico_rkt.py

```diff
@@ -20,381 +20,474 @@
+
+        :return The plugin return code.
+        """
+        rc = 0
+        try:
+            self._execute()
+        except SystemExit, e:
+            # SystemExit indicates an error that was handled earlier
+            # in the stack.  Just set the return code.
+            rc = e.code 
+        except BaseException:
+            # An unexpected Exception has bubbled up - catch it and
+            # log it out.
+            _log.exception("Unhandled Exception killed plugin")
+            rc = 1
+        finally:
+            _log.debug("Execution complete, rc=%s", rc)
+            return rc
+
+    def _execute(self):
+        """Private method to execute this plugin.
+
+        Uses the given CNI_COMMAND to determine which action to take.
+
+        :return: None.
+        """
+        if self.command == CNI_CMD_ADD:
+            # If an add fails, we need to clean up any changes we may
+            # have made.
+            try:
+                self.add()
+            except BaseException:
+                _log.exception("Failed to add container - cleaning up")
+                # TODO - Clean up after a failure.
+                sys.exit(1)
+        else:
+            assert self.command == CNI_CMD_DELETE, \
+                    "Invalid command: %s" % self.command
+            self.delete()
+
+    def add(self):
+        """"Handles CNI_CMD_ADD requests. 
+
+        Configures Calico networking and prints required json to stdout.
+
+        :return: None.
+        """
+        _log.info('Configuring pod %s' % self.container_id)
+
+        # Step 1: Assign an IP address using the given IPAM plugin.
+        assigned_ip = self._assign_ip()
+
+        # Step 2: Create the Calico endpoint object.
+        endpoint = self._create_endpoint(assigned_ip)
+
+        # Step 3: Provision the veth for this endpoint.
+        endpoint = self._provision_veth(endpoint)
+        
+        # Step 4: Provision / set profile on the created endpoint.
+        self._set_profile_on_endpoint(endpoint, self.network_name)
+
+        # Step 5: If all successful, print the IPAM plugin's output to stdout.
+        dump = json.dumps(self.ipam_result)
+        _log.info("Dumping info to container manager: %s", dump)
+        print(dump)
+
+        _log.info("Finished creating pod %s", self.container_id)
+    
+    def delete(self):
+        """Handles CNI_CMD_DELETE requests.
+
+        Remove this pod from Calico networking.
+
+        :return: None.
+        """
+        _log.info('Deleting pod %s' % self.container_id)
+
+        # Step 1: Get the Calico endpoint for this workload. If it does not
+        # exist, log a warning and exit successfully.
+        endpoint = self._get_endpoint()
+
+        # Step 2: The endpoint exists - remove any IP assignments.
+        self._release_ips()
+
+        # Step 3: Delete the veth interface for this endpoint.
+        _log.info("Removing veth for endpoint %s", endpoint.name)
+        netns.remove_veth(endpoint.name)
```

Comment:
We'll want error handling for this (as well as a bunch of other stuff) - happy to merge this to master with the promise of a dedicated error handling PR later (as discussed in person).